src/main/java/org/opensearch/index/codec/customcodecs/CustomCodecService.java

```diff
@@ -49,9 +51,21 @@ public CustomCodecService(MapperService mapperService, IndexSettings indexSettin
         if (mapperService == null) {
             codecs.put(ZSTD_CODEC, new Zstd99Codec(compressionLevel));
             codecs.put(ZSTD_NO_DICT_CODEC, new ZstdNoDict99Codec(compressionLevel));
+            codecs.put(QAT_DEFLATE_CODEC, new QatDeflate99Codec(compressionLevel, () -> {
```

Comment:
Do we know what would happen if the QAT is not supported on the instances, and the user ends up choosing this codec?

Reply:
If QAT is not supported and the `index.codec.qatmode` is `hardware`, an exception will be thrown. If the mode is `auto`, a software execution would happen.

Reply:
I was thinking about that a bit and I strikes me this is very hard constraint: basically what it means, if user makes snapshot of the index, she only will be able to restore it on the compatible hardware (if mode is not `auto`) and, and I am not 100% sure we support changing index codec on restore (we need to double check that, we allow some changes [1])

[1] https://www.elastic.co/guide/en/elasticsearch/reference/7.10/restore-snapshot-api.html#restore-snapshot-api-index-settings

Reply:
@reta I agree, and I don't think we support the change of codec during restores. If the mode is not supported, the shard will fail.

Reply:
@reta @sarthakaggarwal97 The `index.codec.qatmode` has really nothing to do with the index or the codec itself. It's just a way to inform QAT of the user's intention to either use hardware or software. It's a dynamic setting that is read whenever a `newCompressor` or `newDecompressor` instance is [created](https://github.com/mulugetam/custom-codecs/blob/add_qat_acceleration/src/main/java/org/opensearch/index/codec/customcodecs/QatLz4CompressionMode.java#L53). 

Not sure if I missed anything, but are you saying that there are circumstances where executing the below may fail?

```
curl -X PUT "endpoint/<index>" -H 'Content-Type: application/json' -d '
{
  "settings": {
    "index.codec.qatmode": "auto"
  }
}'
```

Reply:
@mulugetam no, I am talking about snapshot / restore. Users could backup (make a snapshot) of the index and than restore it later. 

> If QAT is not supported and the index.codec.qatmode is hardware, an exception will be thrown. If the mode is auto, a software execution would happen.

This would be a hard blocker.

Reply:
@reta, it is possible to override the index settings during a restore -- I just tried it and it works. And, if we change the default `index.codec.qatmode` from `hardware` (current one) to `auto`, then there would be no need to modify the setting during restore. 

The `hardware` mode is included to support uses cases where users prefer hardware-only execution (that is, exception will be thrown if QAT is not available). 
 
 @sarthakaggarwal97 thank you for bringing this up.

Reply:
> @reta, it is possible to override the index settings during a restore -- I just tried it and it works. And, if we change the default `index.codec.qatmode` from `hardware` (current one) to `auto`, then there would be no need to modify the setting during restore.

That is great, thanks you for confirming @mulugetam !